src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackDecoder.cs

```diff
@@ -112,256 +110,278 @@ internal HPackDecoder(int maxDynamicTableSize, int maxResponseHeadersLength, Dyn
-                            // 0 should always be represented by a single byte, so we shouldn't need to check for it in the continuation case.
-                            Debug.Assert(_integerDecoder.Value != 0, "A header name length of 0 should never be encoded with a continuation byte.");
-
-                            OnStringLength(_integerDecoder.Value, nextState: State.HeaderName);
+                            _state = State.HeaderNameIndex;
                         }
-
-                        break;
-                    case State.HeaderName:
-                        _stringOctets[_stringIndex++] = b;
-
-                        if (_stringIndex == _stringLength)
+                    }
+                    else if ((b & DynamicTableSizeUpdateMask) == DynamicTableSizeUpdateRepresentation)
+                    {
+                        // https://tools.ietf.org/html/rfc7541#section-4.2
+                        // This dynamic table size
+                        // update MUST occur at the beginning of the first header block
+                        // following the change to the dynamic table size.
+                        if (_headersObserved)
                         {
-                            OnString(nextState: State.HeaderValueLength);
+                            throw new HPackDecodingException(SR.net_http_hpack_late_dynamic_table_size_update);
                         }
 
-                        break;
-                    case State.HeaderValueLength:
-                        _huffman = (b & HuffmanMask) != 0;
-
-                        if (_integerDecoder.StartDecode((byte)(b & ~HuffmanMask), StringLengthPrefix))
+                        if (_integerDecoder.BeginTryDecode((byte)(b & ~DynamicTableSizeUpdateMask), DynamicTableSizeUpdatePrefix, out intResult))
                         {
-                            if (_integerDecoder.Value > 0)
-                            {
-                                OnStringLength(_integerDecoder.Value, nextState: State.HeaderValue);
-                            }
-                            else
-                            {
-                                OnStringLength(_integerDecoder.Value, nextState: State.Ready);
-                                OnHeaderComplete(onHeader, onHeaderState, new ReadOnlySpan<byte>(_headerName, 0, _headerNameLength), new ReadOnlySpan<byte>());
-                            }
+                            SetDynamicHeaderTableSize(intResult);
                         }
                         else
                         {
-                            _state = State.HeaderValueLengthContinue;
+                            _state = State.DynamicTableSizeUpdate;
                         }
+                    }
+                    else
+                    {
+                        // Can't happen
+                        Debug.Fail("Unreachable code");
+                        throw new InvalidOperationException();
```

Comment:
If it can't happen and it's unreachable code, do we need the `throw` at all?

Reply:
I'd argue yes. If we make it an invariant that this branch can't happen, why not throw an exception? Otherwise, discovering if the invariant is wrong is hard to determine if running in Release.

Reply:
Would it make sense to add a suitable error message in the exception? Preferably a euphemism for "can't happen"?

Reply:
Already does!